src/utils.zig

```diff
@@ -31,35 +33,62 @@ pub const toMagicNumberLittle = switch (native_endian) {
     builtin.Endian.Big => toMagicNumberForeign,
 };
 
-pub fn readStructNative(reader: io.StreamSource.Reader, comptime T: type) !T {
-    return try reader.readStruct(T);
+fn checkEnumFields(data: anytype) error{InvalidEnumTag}!void {
+    const T = @typeInfo(@TypeOf(data)).Pointer.child;
+    inline for (meta.fields(T)) |entry| {
+        switch (@typeInfo(entry.field_type)) {
+            .Enum => {
+                const value = @enumToInt(@field(data, entry.name));
+                _ = try std.meta.intToEnum(entry.field_type, value);
+            },
+            .Struct => {
+                try checkEnumFields(&@field(data, entry.name));
+            },
+            else => {},
+        }
+    }
 }
 
-pub fn readStructForeign(reader: io.StreamSource.Reader, comptime T: type) !T {
-    comptime std.debug.assert(@typeInfo(T).Struct.layout != builtin.TypeInfo.ContainerLayout.Auto);
-
-    var result: T = undefined;
+pub fn readStructNative(reader: io.StreamSource.Reader, comptime T: type) StructReadError!T {
+    var result: T = try reader.readStruct(T);
+    try checkEnumFields(&result);
+    return result;
+}
 
+fn swapFieldBytes(data: anytype) error{InvalidEnumTag}!void {
+    const T = @typeInfo(@TypeOf(data)).Pointer.child;
     inline for (meta.fields(T)) |entry| {
         switch (@typeInfo(entry.field_type)) {
-            .ComptimeInt, .Int => {
-                @field(result, entry.name) = try reader.readIntForeign(entry.field_type);
+            .Int => |int| {
+                if (int.bits > 8) @field(data, entry.name) = @byteSwap(entry.field_type, @field(data, entry.name));
```

Comment:
The code will read better if it will look like this
```zig
if (int.bits > 8) {
    @field(data, entry.name) = @byteSwap(entry.field_type, @field(data, entry.name));
}
```

Not a bloker for merge, but a reminder for next PR :)